File: frontend/src/api/monthly.ts
import { api } from "./client";

export interface MonthlyBucketRow {
  bucket: string;
  actual: number;
  plan: number;
  py_actual: number;
}

export interface MonthlyPeriodTable {
  rows: MonthlyBucketRow[];
  higherme: MonthlyBucketRow | null;
}

export interface MonthlyBasis {
  label: string;
  periods: Record<string, MonthlyPeriodTable>; // mtd | qtd | ytd
  trend_actual: number[];
}

export interface MonthlyRecord {
  month: string; // "2026-06"
  label: string; // "June 2026"
  status: string; // prelim | final
  prepared_at: string;
  source_note: string;
  trend_months: string[];
  trend_plan: number[];
  bases: Record<string, MonthlyBasis>; // amt_annualized | w2_uplift
}

export interface MonthlyIndex {
  months: string[];
  latest: string | null;
}

export function fetchMonthlyIndex(): Promise<MonthlyIndex> {
  return api<MonthlyIndex>("/api/monthly-results");
}

export function fetchMonthlyRecord(month: string): Promise<MonthlyRecord> {
  return api<MonthlyRecord>(
    `/api/monthly-results/${encodeURIComponent(month)}`,
  );
}

File: backend/app/schemas/monthly.py
"""Monthly reported bookings results — the finance-blessed numbers.

A record is one report month: Actual vs Plan vs Prior-Year by Bookings Type
(motion), for MTD/QTD/YTD, on both actuals bases, plus the monthly trend
series. Mirrors the structure of the monthly Bookings Results workbook so a
new month can be loaded as one document.
"""
from __future__ import annotations

from pydantic import BaseModel, Field


class BucketRow(BaseModel):
    bucket: str
    actual: float
    plan: float
    py_actual: float


class PeriodTable(BaseModel):
    rows: list[BucketRow]
    higherme: BucketRow | None = None


class BasisTables(BaseModel):
    label: str
    periods: dict[str, PeriodTable]  # keys: mtd | qtd | ytd
    trend_actual: list[float] = Field(default_factory=list)


class MonthlyRecord(BaseModel):
    month: str  # "2026-06"
    label: str  # "June 2026"
    status: str = "prelim"  # prelim | final
    prepared_at: str = ""
    source_note: str = ""
    trend_months: list[str] = Field(default_factory=list)
    trend_plan: list[float] = Field(default_factory=list)
    bases: dict[str, BasisTables]  # keys: amt_annualized | w2_uplift


class MonthlyIndex(BaseModel):
    months: list[str]
    latest: str | None
